### ISSUE TITLE
Add support for Workflow Model

### ISSUE THREAD

**Issue by sfmskywalker** [elsa-workflows/elsa-core#596]
Currently, a workflow can define variables to store and retrieve state using `SetVariable` and `GetVariable` activities and methods. This works fine, but the syntax can become a bit cumbersome.

What we could have in addition is the notion of a "model" (or data or state) that is just like a variable, but available through a new strongly-typed property called `Model`. This would be somewhat analogous to the `Model` property on a Razor view.

This allows the user to define workflows that look very similar to the way it's done in [Workflow Core](https://github.com/danielgerlag/workflow-core#sample-use-cases):

```csharp
public class MyModel
{
	public string Email { get; set; }
	public string Password { get; set; }
	public string UserId { get; set; }
}

public class MyWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder<MyModel> builder)
    {    
        builder
            .StartWith<CreateUser>()
                 .With(activity=> activity.Email, model=> model.Email)
                 .With(activity=> activity.Password, model => model.Password)
           .Then(context=> context.Model.UserId = context.Input<string>())
           .Then<SendConfirmationEmail>()
                .With(activity => activity.Email, model => model.Email)
                .With(activity => activity.UserId, model => model.UserId)
           .Then<WaitForConfirmation>()
                .With(activity => activity.UserId, model => model.UserId)
           .Then<ActivateUser>()
               .With(activity => activity.UserId, model => model.UserId);
    }
}
```

**Comment by zbogdanov** [comment]
Hi @sfmskywalker! First of all, thanks for all the great work you did so far.

Secondly, I'm currently evaluating Elsa for our own internal needs to power a very extendible and large set of mission-critical workflows. We decided to investigate third-party solutions when we finished the design (on paper) of our own in-house solution, and then comparing the two for requirements.

One of the most useful things for our needs is the "Workflow Context" concept, where as the workflow gets activated and starts to run, a last snapshot of the full context is retrieved from the data store and the rest is generated dynamically - creating one big model to be used by the workflow for evaluation and other needs. The context can get updated as the workflow runs also by set variable activities. Right now Elsa does not have this and this would be one of the first things we would need to extend if we decide to go with Elsa, so this feature is a great idea I think. 

I'll be keeping an eye on this PR and If we decide to go to Elsa, I'll probably volunteer to help out with this feature if it will not be done by then.

Thanks again, keep up the fantastic work!

**Comment by sfmskywalker** [comment]
@zbogdanov Thanks for the kind words!

Just to be sure I understand correctly: Elsa 2 today supports the concept of Workflow Context: it loads and persists workflow contexts by calling out to a configurable workflow context provider.

This means you should already be able to update the workflow context from workflow activities, and the updated workflow context would already be persisted for you (through your own provider).

You can even configure the “fidelity” of when the workflow context gets refreshed and persisted:

- Before each burst of execution (which means: when a workflow starts/resumes)
- Before each activity execution
- Before certain activities execute (configurable per individual activity)

Given the above, what else would you need for your use cases?

To be clear: the Workflow Model described in this issue is a separate feature. It does not store the values external to the workflow; rather, it stored the model as part of the workflow itself, exactly like workflow variables are. It’s really just a more convenient syntax to access and store multiple bits of information.

**Comment by zbogdanov** [comment]
@sfmskywalker Ahh, ok, I'm experimenting with Elsa 1 not 2 so that's why I didn't see this concept. I'm going to switch to elsa 2 to paint a more detailed picture... Was hesitant as I'm not sure on what is the status of the workflow designer but I'll investigate more thoroughly.

In this case I think the workflow context described should be a good fit for us. Thank you!

**Comment by sfmskywalker** [comment]
Makes sense. The designer in Elsa 2 is under construction and not very usable right now. The workflow engine itself should work fine in its current state, so you might consider using the workflow builder API for your experiments. I hope to have the designer finished by the end of Q1.

**Comment by mohdali** [comment]
This should also extend to the designer to allow picking the model parameters as inputs to the activity. This will make it easier to design the workflows, add validation and reduce user errors.

**Comment by craigfowler** [comment]
@sfmskywalker This is very interesting, partly because it touches on something I was planning on discussing after Elsa 2.0.  That's the separation of **State** and **Functionality** within activities.

## My waffling about architectural ideas
Right now anything implementing `IActivity` has to maintain both its own state as well as the methods like `OnExecuteAsync` etc.  To simplify the API I was going to talk about separating those two things into two classes.  There's probably some fringe benefits too, such as making serialization easier and also reducing memory overheads because each specific activity instance is now a stateless service (so we need fewer instances).

What I had in mind was that `IActivity` defines only that the state of an activity is `System.Object` (which could be `null` for a stateless activity) but there would also be a generic `IActivity<TState> where TState : class`.  Developers writing their own activities implement the generic interface and on the activity class they're just writing the functional methods which define how the activity works.  Then, for the activity state they write a small POCO model/state object.

Those functional methods (again, example is `OnExecuteAsync`) - in the generic version of the activity interface - would provide an instance of that model/state as an additional parameter:

```csharp
ValueTask<bool> CanExecuteAsync(ActivityExecutionContext context, TState state);
```

Combined with some other tricks, we might be able to do away with the base `Activity` class altogether (and I'm a strong believer in eliminating class inheritance where possible).  If consumers only need to implement an interface then their code becomes simpler and also easier to unit test (which means more chance that they will actually write them).

## Triage & priority
Anyway - all of that aside - I'm going to mark this up as **do later** because you've not added it to the 2.0 milestone.  I figure that means you intend to revisit this in 3.0 or later.  Feel free to move to **do soon** if you think this should be looked-at earlier.

**Comment by sfmskywalker** [comment]
@craigfowler I like the suggestion and would like to explore this further for some of the 3.0 re-architecture being discussed [here](1193) as well.

**Comment by craigfowler** [comment]
Hello again!

What we have in Elsa 1.x and 2.x is not that dissimilar to the concept of a `ViewBag` (minus `dyanmic` typing) or an untyped model from ASP.NET MVC 5.  The workflow instance has a "state" object which is essentially a dictionary of objects (accessed as workflow "variables").

At the workflow-level I don't think you can ever _get rid of_ that arbitrary state to be honest.  Otherwise, every activity in the workflow which needs workflow-level state storage needs to know about that specific model class.  That would very quickly become impossible to manage and cause more problems than it would solve.

I think it's reasonable to do (again) what ASP.NET MVC5 did though, and allow a strongly-typed model _to supplement_ the arbitrary key/value storage.  Interestingly MVC5 accomplishes this via a reserved "arbitrary variable name" named `Model`, which is then always presented in its strongly-typed manner.  That's a bit of a fudge; perhaps Elsa could do better?

## Is it worth doing?
The follow-up though is "What are the benefits of a strongly-typed model, _at the Workflow level_?" and who would want to use it?  Clearly the only activities which may use such a model are those activities which are written with that specific workflow (and its model) in mind.  Coding such an activity in that manner would limit its reuse outside of those specific workflows though.  An activity which is made for a workflow whose model is a `MyFunkyModel` can't be used on a workflow whose model is a `MyJazzyModel`.

Clearly for maximum reusability, you want your activities to be built just against the arbitrary variables.

## Could someone get the same benefit by saving a model in a variable?
Here's an interesting question/thought (and I don't know what would happen in this scenario).  What if I wanted to share a strongly typed model through a series of related activity classes.  Could the activity not `SetVariable("MyStronglyTypedModel", new MyFunkyModel())` (or similar).  If you wanted to share that model between activities, you'd use the variable name from a constant which is known to all of the activities which want to look at/manipulate that model.

How does the serialization work?  Is the in-memory state (whilst the workflow is executing during a burst of operation) the serialized state?  IE:  `GetVariable`/`SetVariable` serializes/deserializes to-order?  Or is it completely deserialized once to a "plain in-memory store" (like `Dictionary<string,object>` or conceptually similar) at the beginning of the burst and then only re-serialized once the burst of workflow activity completes?

If the latter is true then this would make it quite easy for developers to create their own arbitrary C# strongly-typed models as variable values.  Mutations to that model "would automatically stick" and be serialized.  In other words I could do something like `GetVariable<MyFunkyModel>("MyStronlyTypedModel").BringBackTheFunk = true;` and it would actually update the state of the model and that mutation would be saved with the workflow.

If the former were true, then the trick above wouldn't work, because each time we use `GetVariable`, I'm not getting that single shared object reference from an in-memory state.  Instead, a new copy of that state is being deserialized-to-order for me and mutating it alone wouldn't be good enough.  I would need to subsequently `SetVariable` to "put it back" into the workflow's serialized state, in order to make my changes stick.

Depending on which of those two scenarios above is true, this might be good enough for developers who want to write specific activities which make use of a specific strongly-typed model.

**Comment by sfmskywalker** [comment]
Hi @craigfowler, welcome back!

Just a quick note while I was skimming over your remarks while hiking through the forests:

In my view, none of the activities ever need to be aware of any strongly typed model.

It’s really mostly about the developer experience when they write workflows in code and design them with the designer. It’s about activity input properties whose values are **expressions and lambdas** that would benefit from a strongly typed model.

None of the activities know about this model. They just evaluate the expression (or as is the case with V2 they receive the result of the evaluated expression).

Back to my hike 🚶‍♂️
